lang/funcs/collection.go

```diff
@@ -170,32 +170,43 @@ var CoalesceListFunc = function.New(&function.Spec{
 	Params: []function.Parameter{},
 	VarParam: &function.Parameter{
 		Name:             "vals",
-		Type:             cty.List(cty.DynamicPseudoType),
+		Type:             cty.DynamicPseudoType,
 		AllowUnknown:     true,
 		AllowDynamicType: true,
 		AllowNull:        true,
 	},
 	Type: func(args []cty.Value) (ret cty.Type, err error) {
 		if len(args) == 0 {
-			return cty.NilType, fmt.Errorf("at least one argument is required")
+			return cty.NilType, errors.New("at least one argument is required")
 		}
 
 		argTypes := make([]cty.Type, len(args))
 
 		for i, arg := range args {
+			// if any argument is unknown, we can't be certain know which type we will return
+			if !arg.IsKnown() {
+				return cty.DynamicPseudoType, nil
+			}
+			ty := arg.Type()
+
+			if !ty.IsListType() && !ty.IsTupleType() {
+				return cty.NilType, errors.New("coalescelist arguments must be lists or tuples")
```

Comment:
This constraint is fine for now I think, but it's interesting to note that this function could reasonably be defined for anything that has a length, which would include sets. Since this is an error rather than an automatic conversion though, we can safely add support for more types later if we decide to, so good to be conservative for now and see what comes up in real-world use.

Reply:
The reason I left this one, is because it's documented to work with "lists", and returns one of the list arguments (plus, it's also called `coalescelist` ;) ). I think I'd rather modify Coalesce if we want something that handles multiple types.